diff --git a/src/crypto/recipient.cpp b/src/crypto/recipient.cpp
index 0a3bc1cdb..eb0f5baa9 100644
--- a/src/crypto/recipient.cpp
+++ b/src/crypto/recipient.cpp
@@ -87,7 +87,7 @@ Recipient::Recipient(const Mailbox &mb)
 
 void Recipient::detach()
 {
-    if (d && !d.unique()) {
+    if (d && d.use_count() != 1) {
         d.reset(new Private(*d));
     }
 }
diff --git a/src/crypto/sender.cpp b/src/crypto/sender.cpp
index 1e3cd53bf..5be3d5dd9 100644
--- a/src/crypto/sender.cpp
+++ b/src/crypto/sender.cpp
@@ -93,7 +93,7 @@ Sender::Sender(const Mailbox &mb)
 
 void Sender::detach()
 {
-    if (d && !d.unique()) {
+    if (d && d.use_count() != 1) {
         d.reset(new Private(*d));
     }
 }
-- 
GitLab

