--- a/CMakeLists.txt	2021-11-10 02:25:40.000000000 -0800
+++ b/CMakeLists.txt	2022-06-20 10:45:04.075301008 -0700
@@ -6,6 +6,7 @@
 set(ADWAITAQT_VERSION "1.4.1")
 
 option(USE_QT6 "Use Qt6 instead of Qt5" OFF)
+option(USE_XCB "Use XCB Bindings for Qt5" ON)
 
 if (USE_QT6)
     set(QT_MIN_VERSION "6.2.0")
@@ -38,7 +39,7 @@
     Widgets
 )
 
-if (NOT APPLE AND NOT WIN32 AND NOT USE_QT6)
+if (NOT APPLE AND NOT WIN32 AND NOT USE_QT6 AND USE_XCB)
     find_package(XCB 1.10 COMPONENTS XCB)
 
     set(ADWAITA_HAVE_X11 ${XCB_FOUND})
