--- a/meson_options.txt	2021-02-22 08:36:29.000000000 -0800
+++ b/meson_options.txt	2021-12-30 03:41:47.433110716 -0800
@@ -14,5 +14,6 @@
 option('tests', type: 'boolean', value: false, description: 'Build tests')
 option('introspection', type: 'boolean', value: true, description: 'Enable introspection for this build')
 
+option('duktape', type: 'boolean', value: true, description: 'use duktape instead of mozjs')
 option('gtk_doc', type: 'boolean', value: false, description: 'use gtk-doc to build documentation')
 option('man', type: 'boolean', value: false, description: 'build manual pages')
--- a/meson.build	2021-09-29 12:16:29.000000000 -0700
+++ b/meson.build	2021-12-30 03:49:42.911694250 -0800
@@ -133,8 +133,13 @@
 assert(cc.has_header('expat.h', dependencies: expat_dep), 'Can\'t find expat.h. Please install expat.')
 assert(cc.has_function('XML_ParserCreate', dependencies: expat_dep), 'Can\'t find expat library. Please install expat.')
 
-mozjs_dep = dependency('mozjs-78')
-
+duktape = get_option('duktape')
+if duktape
+  mozjs_dep = dependency('duktape', version: '>= 2.0')
+  libm_dep = cc.find_library('m')
+else
+  mozjs_dep = dependency('mozjs-78')
+endif
 dbus_dep = dependency('dbus-1', required: false)
 dbus_policydir = pk_prefix / pk_datadir / 'dbus-1/system.d'
 if dbus_dep.found()
--- a/src/polkitbackend/meson.build	2021-07-29 08:16:35.000000000 -0700
+++ b/src/polkitbackend/meson.build	2021-12-30 04:36:06.872865120 -0800
@@ -5,7 +5,6 @@
   'polkitbackendactionpool.c',
   'polkitbackendauthority.c',
   'polkitbackendinteractiveauthority.c',
-  'polkitbackendjsauthority.cpp',
 )
 
 output = 'initjs.h'
@@ -31,6 +30,13 @@
   '-DPACKAGE_SYSCONF_DIR="@0@"'.format(pk_prefix / pk_sysconfdir),
 ]
 
+if duktape
+  sources += files('polkitbackendduktapeauthority.c')
+  deps += libm_dep
+else
+  sources += files('polkitbackendjsauthority.cpp')
+endif
+
 if enable_logind
   sources += files('polkitbackendsessionmonitor-systemd.c')
 
