diff -ru two/musique.pro flaviotordini_musique/musique.pro
--- two/musique.pro	2016-12-09 00:05:33.857768933 -0800
+++ flaviotordini_musique/musique.pro	2016-12-09 01:07:17.500868113 -0800
@@ -16,9 +16,10 @@

 TARGET = $${APP_UNIX_NAME}
 
-QT += network sql widgets
+QT += network sql widgets
+
+CONFIG += qtsingleapplication phonon4qt5

-include(src/qtsingleapplication/qtsingleapplication.pri)
 include(src/http/http.pri)
 include(src/tags/tags.pri)

@@ -149,7 +150,7 @@
     src/httputils.cpp \
     src/tagchecker.cpp
 RESOURCES += resources.qrc
-DESTDIR = build/target/
+isEmpty(DESTDIR):DESTDIR = build/target/
 OBJECTS_DIR = build/obj/
 MOC_DIR = build/moc/
 RCC_DIR = build/rcc/ 
